cargoSetupHook, rustc: Force frame pointers for !isx86_32

dramforever opened the pull request:
We did this for C in #399014 but not Rust. Add this to cargoSetupHook.

Also add to rustc because the bootstrap doesn't use cargoSetupHook.

To test:

```nix
nix build -f . rustc cargo pkgsCross.riscv64.pay-respects
objdump -j .text -d ...

# Get this from pkgsCross.riscv64.stdenv.cc
riscv64-unknown-linux-gnu-objdump -j .text -d ...
```

(There may be wrapper scripts that you need to get behind)

Patterns to look for to confirm usage of frame pointers

```
# x86_64

000000000040aa00 <_ZL28find_plugindir_spec_functioniPPKc>:
  40aa00:       55                      push   %rbp
  40aa01:       48 89 e5                mov    %rsp,%rbp  # <- Good
```

```
# aarch64

000000000016a3a4 <_ZN3std3sys12thread_local6native4lazy20Storage$LT$T$C$D$GT$10initialize17h53f3010de99a7b15E>:
  16a3a4:       d10143ff        sub     sp, sp, #0x50
  16a3a8:       a9037bfd        stp     x29, x30, [sp, #48] # <- Note x29, x30 and number
  16a3ac:       f90023f3        str     x19, [sp, #64]
  16a3b0:       9100c3fd        add     x29, sp, #0x30      # <- Good, number same as above but hex
```

```
# riscv64

000000000000e5ee <_ZN3std3sys12thread_local6native4lazy20Storage$LT$T$C$D$GT$10initialize17h151587489d666680E>:
    e5ee:       1141                    addi    sp,sp,-16 # <- Note this number
    e5f0:       e406                    sd      ra,8(sp)
    e5f2:       e022                    sd      s0,0(sp)
    e5f4:       0800                    addi    s0,sp,16  # <- Good, number same as above but positive
```

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dramforever committed:
cargoSetupHook: Force frame pointers for !isx86_32

We did this for C in #399014 but not Rust. Add
-Cforce-frame-pointers=yes to rustflags in cargoSetupHook

dtomvan commented:
Could you maybe share the error you are encountering with rustc? Maybe the stage0 compiler doesn't like getting compiled using frame pointers...

dramforever commented:
@dtomvan Perhaps there was some confusion. There was no error, it compiled perfectly fine, and the generated code does not use frame pointers.

I checked on aarch64-linux, and will try x86_64-linux later just in case it was a fluke.

dramforever commented:
To reproduce:

- rebase this back on master so it's actually reasonable to build
- `nix build -f . rustc.unwrapped`
- `objdump -d` it
- Skip the `.plt`, go to `.text`
- Check assembly codd

Expected: frame pointers are used. Actual: doesn't seem to be used.

dtomvan commented:
I'll give it a shot on `x86_64-linux`.

dramforever commented:
I'm looking for this on function entry:

```
000000000040aa00 <_ZL28find_plugindir_spec_functioniPPKc>:
  40aa00:       55                      push   %rbp
  40aa01:       48 89 e5                mov    %rsp,%rbp  # <- good
```

I found those on x86_64 gcc. If you try e.g. x86_64 cargo, I think you will also get these. They don't have to be exactly the first two instructions.

It's also normal for some functions to not have this. For example, a lot of `.cold` functions don't have this, which makes sense because they're not really functions in their own rights, but chunks of other functions moved out of the way.

However, in the rustc disassembly provided, a large amount of functions don't seem to have this problem, for some reason. It seems that only functions from libstd were given frame pointers, whereas rustc's own functions are not.

alyssais commented:
Do we definitely want to do this in the wrapper, rather than the build system? The existing flag set in the wrapper is a bit of a special case where it was the only feasible way to do it…

dramforever commented:
I originally thought setting this here would make it work for compiling rustc as well, but it seems to be not so easy...

dramforever committed:
rustc: Set rust.frame-pointers if !isx86_32

We want -Cforce-frame-pointers=yes when building rustc itself, but the
bootstrap process does not use cargoSetupHook, so set the bootstrap
config option rust.frame-pointers = true if target is not x86_32.

Note that std is unconditionally compiled with frame pointers. See code
comment for elaboration.

See also #399014 for discussion when we enabled frame pointers for C.

dramforever commented on pkgs/build-support/rust/hooks/default.nix:
```diff
@@ -91,13 +91,21 @@
           [target."${stdenv.hostPlatform.rust.rustcTarget}"]
           "linker" = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc"
+          "rustflags" = [ ${
+            lib.optionalString (!stdenv.hostPlatform.isx86_32) ''"-Cforce-frame-pointers=yes"''
```
Firstly, I just did it for consistency. Secondly, we do need to set this for non-cross, but the part above for target is only included for cross, so I thought it ends up clearer this way.

If you specifically think we should *avoid* setting it for host on cross then I can change it.

alyssais replied:
Nope, that makes sense.

dramforever commented on lib/systems/default.nix:
```diff
@@ -522,7 +522,7 @@ let
             # https://github.com/rust-lang/cargo/pull/9169
             # https://github.com/rust-lang/cargo/issues/8285#issuecomment-634202431
-            cargoEnvVarTarget = replaceString "-" "_" (toUpper final.rust.cargoShortTarget);
+            cargoEnvVarTarget = replaceString "-" "_" (toUpper final.rust.rustcTarget);
```
Should this be in a separate PR? It doesn't really change anything except fixes eval for (afaict only) rust-hypervisor-firmware. I thought it would make sense to just do this in one PR.

alyssais replied:
Have you tested this beyond eval? I'm not sure it's right. Looking at Cargo's code it looks a lot like the name used for configuration (which can be set in the environment) is derived from the JSON path:

https://github.com/rust-lang/cargo/blob/930b4f62cfcd1f0eabdb30a56d91bf6844b739bf/src/cargo/core/compiler/compile_kind.rs#L192-L209

alyssais replied:
(I think it would make more sense to be in a separate PR. It doesn't seem to be particularly related to frame pointers.)

dramforever replied:
Hmm, I thought it was an obvious typo.

Without this change, eval fails for `rust-hypervisor-firmware` with this very funny error message: https://github.com/NixOS/nixpkgs/actions/runs/15958033841/job/45006880188#step:5:955

```
       error: the string 'CARGO_TARGET_IIHHC9567HBMDYIJQY88H56SRH5Q9H30_X86_64_UNKNOWN_NONE_RUSTFLAGS' is not allowed to refer to a store path (such as 'iihhc9567hbmdyijqy88h56srh5q9h30-x86_64-unknown-none.json')
```

This is ... not how cargo environment variables work.

What was supposed to happen is the key `target.x86_64-unknown-none.rustflags` is uppercased and underscored, and prepended with `CARGO_`. The right way to set target specific config options in `.cargo/config.toml` is `[target."${stdenv.*Platform.rust.rustcTarget}"]`, so the same goes for `CARGO_TARGET_*` environment variables.

I'll move this change to a separate PR though.

dramforever replied:
I guess this means the `.cargo/config.toml` generation logic is *also* wrong then?

alyssais replied:
> Putting the spec file inside a directory to hide the hash from Nix would probably help.

builtins.unsafeDiscardStringContext would also be fine here I think.

dramforever replied:
So, after reading every occurrence of the word `rustcTarget` in nixpkgs, I came to the conclusion that we should put the JSON into a directory to minimize surprises.

For example, the current `doc/languages-frameworks/rust.section.md` says

> - To pass a completely custom target, define `stdenv.hostPlatform.rust.rustcTarget` with its name, and `stdenv.hostPlatform.rust.platform` with the value.  The value will be serialized to JSON in a file called `${stdenv.hostPlatform.rust.rustcTarget}.json`, and the path of that file will be used instead.

However, this is a lie. The file is actually called `{hashpart}-${stdenv.hostPlatform.rust.rustcTarget}.json`.

This not only gives us the headache that is `CARGO_TARGET_IIHHC956...`, but also the problem that when building a crate requiring a custom target, the user can no longer predictably set target-specific cargo config and crate config options for that target. This doesn't seem to be intended.

Indeed, rust-hypervisor-firmware's integration tests assumes that the target json file name is just `{cpu}-unknown-none`: <https://github.com/cloud-hypervisor/rust-hypervisor-firmware/blob/main/src/integration.rs#L463>. Hypothetically, if we were to enable integration tests for r-h-f, we would need to patch these assumptions out as well. There's no reason to believe that other users of `rust.platform` are not making these kinds of (IMO completely reasonable) assumptions.

alyssais commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -105,8 +111,20 @@ stdenv.mkDerivation (finalAttrs: {
+    else
+      null
+  } =
+    rustFlagsFor stdenv.targetPlatform;
```
If the `rust.framepointers` config option isn't working for us we should probably file a bug report rather than just working around it like this.

dramforever replied:
wait, i remembered, i didn't do that because i didn't think it handles the cross case right, but i'll check for sure

dramforever replied:
`rust.frame-pointers` does work, and integrates well with `fastCross`. I will change it to use that.

dramforever commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -260,6 +260,9 @@ stdenv.mkDerivation (finalAttrs: {
       "--enable-use-libcxx"
+    ]
+    ++ optionals (!stdenv.targetPlatform.isx86_32) [
+      "--set rust.frame-pointers"
```
(`L` means localSystem, `C` means crossSystem, `(b, h, t)` means `(build = b, host = h, target = t)`)

This option cannot be set per target. However, for a rustc build, due to the way `fastCross` is designed, this works out just fine:

- The compiler is compiled as (L, L, L), which means `rust.frame-pointers` on rustc itself according to L
- `std` is compiled as `(L, L, C)`, which means `rust.frame-pointers` on std is set according to C
- Even if we go a step further and do a slow cross `(L, C, C)`, both rustc and std has `rust.frame-pointers` set according to C

I should probably add an option to explain that this depends on the way `fastCross` works

dramforever replied:
Oh no, it seems I misunderstood again.

dramforever replied:
Oh no, it seems I misunderstood again. Before https://github.com/rust-lang/rust/pull/141800, the rust std is unconditionally compiled with `-Cforce-frame-pointers=yes`. After (not yet released, targets 1.89), it's ``-Cforce-frame-pointers=yes`

Thanks for the catch! I will change it to hostPlatform and add a comment explaining that it cannot be changed for targetPlatform when building std.

FWIW, I think this is a sane decision. Most of rust's std contains generic functions whose codegen is done when building user crates. For those, force enabling frame pointers in std allows frame pointer based backtraces without having to recompile. Given that their goals align with us here, we'll let rustc's bootstrap process handle it.

alyssais commented on pkgs/development/compilers/rust/rustc.nix:
```diff
@@ -260,6 +260,17 @@ stdenv.mkDerivation (finalAttrs: {
+      #
+      # Note that when compiling std, frame pointers (or at least leaf frame
+      # pointers, depending on version) are unconditionally enabled and cannot
+      # be overriden, so we do not touch that. (Also note this only applies to
```
```suggestion
      # be overridden, so we do not touch that. (Also note this only applies to
```

alyssais reviewed (changes requested):
Looking good! Just a typo left. Do you feel like doing this for Meson too? It's used to compile Rust code in e.g. Mesa. (Can be a separate PR if you want; no need to hold this one up.)

alyssais commented:
For Meson it would definitely have to go through staging.

dramforever commented:
(That was another comment typo fix: "leaf" -> "non-leaf")
